### PR TITLE
Added testing chapter to contact us tutorial

### PR DESCRIPTION
**Pull request by fklosowski**

### CODE REVIEW

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+  description: Follow these step-by-step tutorials to build a Contact Us form on platformOS
+converter: markdown
+---
+
+{% # TODO %}
+At this point it would be good idea to add some automatic testing to our `contacts/create` command validation mechanism, so we do not need to check it manually over and over again.
+
+## Introduction to platformOS Tests Module
+
+The platformOS Tests Module provides a way to write and run unit tests using Liquid.
```
Comment (Slashek) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
The [platformOS Tests Module](https://github.com/Platform-OS/pos-module-tests) provides a way to write and run unit tests using Liquid.
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+| **invalid_object**      | Ensures `object.valid != true`               | ` function contract = 'modules/tests/assertions/invalid_object', contract: contract, object: invalid_form, field_name: 'form' ` |
+| **not_valid_object**    | Alias for `invalid_object`                   | ` function contract = 'modules/tests/assertions/not_valid_object', contract: contract, object: invalid_data, field_name: 'data' ` |
+| **object_contains_object** | Checks if object contains key-value pairs | ` function contract = 'modules/tests/assertions/object_contains_object', contract: contract, given: profile, object_contains: '{"role": "admin"}' ` |
+| **not_presence**        | Ensures field is blank                       | ` function contract = 'modules/tests/assertions/not_presence', contract: contract, object: data, field_name: 'legacy_field' ` |
+
+All assertions receive and return a contract object that collects errors and tracks the number of checks performed. If a condition is not met, an error is registered in the contract.
```
Comment (Slashek) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: If we want, we can also link to the source code - https://github.com/Platform-OS/pos-module-tests/tree/main/modules/tests/public/lib/assertions

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+### Key Features
+
+- Run unit tests written in Liquid.
+- Validate data structures, function results, and error messages.
+- Simulate function calls and check results with built-in assertions.
+- Group multiple tests into a single file.
+
+### Working with tests module
+
+All test files must be placed under `app/lib/tests/`.
```
Comment (Slashek) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: `modules/<module name>/public/lib/tests` is also fine - in general we recommend placing them in `lib/tests`

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+
+  function contract = 'modules/tests/assertions/not_valid_object', contract: contract, object: contact, field_name: 'contact'
+  function contract = 'modules/tests/assertions/presence', contract: contract, object: contact.errors, field_name: 'body'
+%}
+
+{% comment %}
+  Test: Too long body - should be invalid
+{% endcomment %}
+{% liquid
+  assign long_body = ''
```
Comment (Slashek) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: you can also use `assign long_body = 201 | random_string` :) https://documentation.platformos.com/api-reference/liquid/platformos-filters#random-string

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+  for i in (1..201)
+    assign long_body = long_body | append: 'x'
+  endfor
+
+  assign data = '{ "email": "user@example.com", "body": "' | append: long_body | append: '" }' | parse_json
```
Comment (Slashek) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: for readability, you can consider using
```
{% parse_json data %}
  { 
    "email": "user@example.com", 
  "body":  "{{ 201 | random_string }}" 
  }
{% endparse_json %}
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 6]
```diff
@@ -0,0 +1,521 @@
+---
+metadata:
+  title: Testing validation logic in contacts/create command
+  description: Follow these step-by-step tutorials to build a Contact Us form on platformOS
+converter: markdown
+---
+
+{% # TODO %}
+At this point it would be good idea to add some automatic testing to our `contacts/create` command validation mechanism, so we do not need to check it manually over and over again.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
At this point, it's a good idea to add automated tests for the 'contacts/create' command's validation logic, so we don't have to check it manually each time.
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 7]
```diff
@@ -0,0 +1,521 @@
+---
+metadata:
+  title: Testing validation logic in contacts/create command
+  description: Follow these step-by-step tutorials to build a Contact Us form on platformOS
+converter: markdown
+---
+
+{% # TODO %}
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid:8]: Is something missing from here?

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 8]
```diff
@@ -0,0 +1,521 @@
...
+It is designed for testing commands, validations, and even email-sending behavior in a modular and repeatable way.
+
+### Key Features
+
+- Run unit tests written in Liquid.
+- Validate data structures, function results, and error messages.
+- Simulate function calls and check results with built-in assertions.
+- Group multiple tests into a single file.
+
+### Working with tests module
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
### Working with Tests Module
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+| **not_valid_object**    | Alias for `invalid_object`                   | ` function contract = 'modules/tests/assertions/not_valid_object', contract: contract, object: invalid_data, field_name: 'data' ` |
+| **object_contains_object** | Checks if object contains key-value pairs | ` function contract = 'modules/tests/assertions/object_contains_object', contract: contract, given: profile, object_contains: '{"role": "admin"}' ` |
+| **not_presence**        | Ensures field is blank                       | ` function contract = 'modules/tests/assertions/not_presence', contract: contract, object: data, field_name: 'legacy_field' ` |
+
+All assertions receive and return a contract object that collects errors and tracks the number of checks performed. If a condition is not met, an error is registered in the contract.
+
+## Automatic testing
+
+We will check the behavior of `commands/contacts/create` to ensure the command properly validates user input for a contact form.
+Our tests will covers both valid and invalid inputs using different assertion types.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
Our tests will cover both valid and invalid inputs using various assertion types.
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 10]
```diff
@@ -0,0 +1,521 @@
...
+
+### Installing tests module
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
### Install Tests Module
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 11]
```diff
@@ -0,0 +1,521 @@
...
+
+Let's start by installing the test module on our instance. Ensure that you are in the project root, then:
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
Let's start by installing the Test Module on our instance. Make sure you are in the project root, then run:
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 12]
```diff
@@ -0,0 +1,521 @@
...
+
+<pre class="command-line" data-user="user" data-host="host"><code class="language-bash">
+pos-cli modules install tests
+pos-cli modules download tests
+</code></pre>
+
+Deploy to the instance if needed.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
Deploy to your instance if needed:
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 13]
```diff
@@ -0,0 +1,521 @@
...
+
+<pre class="command-line" data-user="user" data-host="host"><code class="language-bash">
+pos-cli deploy staging
+</code></pre>
+
+Now, visit project page URL `/_tests` to see the test runner GUI:
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
Now, visit the project page URL `/_tests` to open the test runner GUI:
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 14]
```diff
@@ -0,0 +1,521 @@
...
+
+{% # TODO: check size %}
+<img loading="lazy" class="w-full" src="{{ 'images/get-started/contact-us-tutorial/contactus_pos_tests_module_ui.png' | asset_url }}" alt="PlatformOS tests module GUI with example test">
+
+Although the example test doesn't contain any assertions and always passes, you can run it to see how the test runner acts.
+
+### Create a directory for tests
+
+Lets create a directory for tests within the app directory:
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
Let's create a directory for your tests within the 'app' directory:
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 15]
```diff
@@ -0,0 +1,521 @@
...
+<pre class="command-line" data-user="user" data-host="host"><code class="language-bash">
+mkdir -p app/lib/tests
+</code></pre>
+
+### Adding test file - happy path scenario
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
### Add a test file - happy path scenario
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 16]
```diff
@@ -0,0 +1,521 @@
...
+
+Create a `contact_create_happy_path_test.liquid` file in `app/lib/tests/` and add the following code:
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
Create a file named `contact_create_happy_path_test.liquid` in `app/lib/tests/`, and add the following code:
```

Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
This test covers the happy path - a case where the user input is valid and everything should work as expected. By verifying that valid input produces a correctly processed contact object, we create a baseline to confirm that the core logic functions properly before testing failure cases.

To do that, create a `contact_create_happy_path_test.liquid` file in `app/lib/tests/` and add the following code:
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 17]
```diff
@@ -0,0 +1,521 @@
...
+
+#### 6. Final Output: Debug Print
+
+{% raw %}
+```liquid
+<p>Contact object: {{ contact }}</p>
+```
+{% endraw %}
+
+This is not part of the test but helps visually debug the contact object after processing — especially useful during test development.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
This is not part of the test logic itself, but it helps visually debug the `contact` object after processing — especially useful during test development.
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 18]
```diff
@@ -0,0 +1,521 @@
...
+Now, visit project page URL `/_tests` and run `tests/contact_create_happy_path_test`.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
#### Run happy path test file

Now, visit your project page at `/_tests` and run the `tests/contact_create_happy_path_test`.
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 19]
```diff
@@ -0,0 +1,521 @@
...
+The test result screen shows debug output we configured (which can be removed in the final version), along with a summary of passed and failed assertions and the execution time.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
The test result screen will display:
- The debug output we configured (which can be removed in the final version),
- A summary of passed and failed assertions,
- And the execution time.
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 20]
```diff
@@ -0,0 +1,521 @@
...
+
+{% # TODO: check size %}
+<img loading="lazy" class="w-full" src="{{ 'images/get-started/contact-us-tutorial/contactus_pos_tests_module_ui_run_happy_path_tests.png' | asset_url }}" alt="PlatformOS tests module GUI with happy path test">
+
+Before moving further, take a moment to play around with our tests — for example, change the test data or switch assertions — to get familiar with the tests module and better understand how it processes different scenarios.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
Before moving on, take a moment to experiment with the test file — for example, change the test data or try different assertions. This will help you become more familiar with the Tests Module and how it handles different scenarios.
```

Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: @fklosowski This is the first point where I got confused and lost a bit: how and where to change the test data and how to switch assertions?

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 21]
```diff
@@ -0,0 +1,521 @@
...
+
+{% render 'alert/note', content:'Data Persistence: These are unit tests – they do not write to the database. The `commands/contacts/create` command only performs formatting and validation. It does not save data via GraphQL mutation (record_create) unless the execute.liquid file is called explicitly, which these tests intentionally avoid to keep tests isolated and fast.'%}
+
+### Adding negative path scenarios
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 22]
```diff
@@ -0,0 +1,521 @@
...
+
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid:258]:
```suggestion
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 23]
```diff
@@ -0,0 +1,521 @@
...
+Now, we would like to add some negative scenarios to our test suite to check whether our command correctly handles various possible cases of invalid input.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 24]
```diff
@@ -0,0 +1,521 @@
...
+
+### Adding test file - negative path
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
### Add a test file - negative path scenarios

Now, let’s add some negative test cases to our suite to verify that the 'contacts/create' command correctly handles various types of invalid input.

```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 25]
```diff
@@ -0,0 +1,521 @@
...
+
+Create a `contact_create_negative_path_test.liquid` file in `app/lib/tests/` and add the following code:
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
Create a file named `contact_create_negative_path_test.liquid` in `app/lib/tests/` and add the following code:
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 26]
```diff
@@ -0,0 +1,521 @@
...
+  function contact = 'commands/contacts/create/build', object: data
+
+  if contact.email != 'user@example.com'
+    function contract = 'modules/tests/helpers/register_error', contract: contract, field_name: 'email', message: 'Email was not trimmed or downcased'
+  endif
+%}
+```
+{% endraw %}
+
+Then run our negative path tests again.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
Now run the negative path tests again.
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 27]
```diff
@@ -0,0 +1,521 @@
...
+
+{% # TODO: check size %}
+<img loading="lazy" class="w-full" src="{{ 'images/get-started/contact-us-tutorial/contactus_pos_tests_module_ui_run_negative_path_tests_fail.png' | asset_url }}" alt="PlatformOS tests module GUI with negative path tests - failure">
+
+Looks like the test we just added failed. If you were paying close attention, you might already know why.
+Based on our current implementation, `create/build.liquid` downcases the email, but it doesn’t trim it yet.
+The test, however, expects the email to be both converted to lowercase and trimmed of any extra spaces.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
However, this test expects the email to be both downcased and trimmed of any extra spaces. Adding extra spaces is a common issue, so this reflects a real-world user behavior, when users are copying/pasting email addresses with extra spaces.
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 28]
```diff
@@ -0,0 +1,521 @@
...
+Adding extra spaces is a common issue—especially when users copy and paste their email—so it might be a good idea to update `build.liquid` to handle trimming as well.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
To fix this, let’s update 'build.liquid' to handle trimming as well.
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 29]
```diff
@@ -0,0 +1,521 @@
...
+
+{% render 'alert/note', content:'This test does not use assertion but extra logic and helper function to register an error, so the number of assertions in test report is still 16. This shows that flexible test logic can be written using Liquid’s native conditionals, helpful for edge cases or compound logic.'%}
+
+First, let's add debugging line at the very end of `contact_create_negative_path_test` and run tests again:
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
First, let's add a debugging line at the very end of `contact_create_negative_path_test` and run tests again:
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 30]
```diff
@@ -0,0 +1,521 @@
...
+
+ {% raw %}
+ ```liquid
+ <p>Contact object: {{ contact }}</p>
+ ```
+ {% endraw %}
+
+Notice that we can take a shortcut and catch the output of our failing test this way only because it is the last scenario.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: For me, the test failed here as well.
I don't entirely understand this statement, can you please elaborate or rephrase it? What does last scenario mean?

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 31]
```diff
@@ -0,0 +1,521 @@
...
+
+You should now see debug output like this:
+
+```json
+{"email":" user@example.com ","body":"Nice body here"}
+```
+
+As you can expect email was downcased but still has spaces around it so we need to implement trimm functionality to meet our new requirement.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
As you can expect, the email has been downcased, but it still has spaces around it— so we’ll need to implement trimming to meet our new requirement.
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 32]
```diff
@@ -0,0 +1,521 @@
...
+
+Alternate `build.liquid` code as follows:
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
Modify the `build.liquid` code as follows:
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 33]
```diff
@@ -0,0 +1,521 @@
...
+Notice that we can take a shortcut and catch the output of our failing test this way only because it is the last scenario.
+
+You should now see debug output like this:
+
+```json
+{"email":" user@example.com ","body":"Nice body here"}
+```
+
+As you can expect email was downcased but still has spaces around it so we need to implement trimm functionality to meet our new requirement.
+
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid:506]:
```suggestion
#### Fix the 'build.liquid' logic

```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 34]
```diff
@@ -0,0 +1,521 @@
...
+  }
+{% endparse_json %}
+
+{% liquid
+  return contact
+%}
+```
+{% endraw %}
+
+We applied `strip` next to `downcase`.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
Here, we’ve added the 'strip' filter alongside 'downcase', ensuring that leading and trailing spaces are removed before the email is normalized.
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 35]
```diff
@@ -0,0 +1,521 @@
...
+Now let's run the tests again. This time, the debug output should look like this:
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
Now run the tests again. This time, the debug output should look like:
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 36]
```diff
@@ -0,0 +1,521 @@
...
+
+```json
+{"email":"user@example.com","body":"Nice body here"}
+```
+So the email is downcased and trimmed as well and all the tests passed.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
With that, the email is now properly downcased and trimmed — and all the tests pass.
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 37]
```diff
@@ -0,0 +1,521 @@
...
+
+## Summary
+
+In this chapter, you learned how to write unit tests for the `contacts/create command` using the platformOS Tests Module, covering both valid and invalid input scenarios with a variety of assertions.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
In this chapter, you learned how to write unit tests for the `contacts/create command` using the platformOS Tests Module. We covered both valid and invalid input scenarios using a variety of assertion types.
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 38]
```diff
@@ -0,0 +1,521 @@
...
+
+Also the `Email is Downcased and Trimmed` test case demonstrated how to test and evolve application logic based on realistic user input patterns—in this case, ensuring an email address is properly normalized by trimming and downcasing it.
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
Also the `Email is Downcased and Trimmed` test case also showed how you can evolve your application logic by writing tests that reflect real-world user input—such as handling extra spaces and inconsistent email casing.
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 39]
```diff
@@ -0,0 +1,521 @@
...
+
+- Small improvements can be enforced through testing.
+
+- Missing requirements can be detected via test failures.
+
+- Correctness is confirmed by rerunning the tests.
+
+- This reinforces a test-first (or test-guided) mindset.
+
+This step encapsulates a powerful development workflow: write a test → see it fail → fix the logic → see it pass—a cornerstone of Test-Driven Development (TDD).
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
It also demonstates a powerful development workflow:
**Write a test → see it fail → fix the logic → see it pass.** 
This cycle is the cornerstone of Test-Driven Development (TDD).
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 40]
```diff
@@ -0,0 +1,521 @@
...
+Lets create a directory for tests within the app directory:
+<pre class="command-line" data-user="user" data-host="host"><code class="language-bash">
+mkdir -p app/lib/tests
+</code></pre>
+
+### Adding test file - happy path scenario
+
+Create a `contact_create_happy_path_test.liquid` file in `app/lib/tests/` and add the following code:
+
+**app/lib/tests/contact_create_happy_path_test.liquid**
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
#### app/lib/tests/contact_create_happy_path_test.liquid
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 41]
```diff
@@ -0,0 +1,521 @@
...
+
+### Adding negative path scenarios
+
+Now, we would like to add some negative scenarios to our test suite to check whether our command correctly handles various possible cases of invalid input.
+
+### Adding test file - negative path
+
+Create a `contact_create_negative_path_test.liquid` file in `app/lib/tests/` and add the following code:
+
+**app/lib/tests/contact_create_negative_path_test.liquid**
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
#### app/lib/tests/contact_create_negative_path_test.liquid
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 42]
```diff
@@ -0,0 +1,521 @@
...
+
+- Ensures errors.email and errors.body are populated
+
+We've tested `contacts/create command` against multiple invalid input scenarios. You can now run the negative path tests to verify that all validation rules are correctly enforced.
+
+### Handling Subtle Input Variations
+
+Now, please add one more scenario at the end of our negative path test file:
+
+**app/lib/tests/contact_create_negative_path_test.liquid**
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
#### app/lib/tests/contact_create_negative_path_test.liquid
```

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid [hunk 43]
```diff
@@ -0,0 +1,521 @@
...
+
+```json
+{"email":" user@example.com ","body":"Nice body here"}
+```
+
+As you can expect email was downcased but still has spaces around it so we need to implement trimm functionality to meet our new requirement.
+
+Alternate `build.liquid` code as follows:
+
+**app/lib/commands/contacts/create/build.liquid**
```
Comment (simplykatsa) [inline app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]:
```suggestion
#### app/lib/commands/contacts/create/build.liquid
```

**Review by simplykatsa — CHANGES_REQUESTED** [review]

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid (hunk 7 above)
Comment (fklosowski) [reply app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid:8]: I will remove it :)

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid (hunk 20 above)
Comment (fklosowski) [reply app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: My intention was for the user to change

from:

`assign data = '{ "email": "USER@example.com", "body": "This is a valid message body." }' | parse_json`

to, for example:

`assign data = '{ "email": "USER@example.com", "body": "Too short body." }' | parse_json`

or to modify the assertion,

from:

```
# 4: Assert the contact.valid is true
function contract = 'modules/tests/assertions/true', contract: contract, object: contact, field_name: 'valid'
```
to:

```
# 4: Assert the contact.valid is true
function contract = 'modules/tests/assertions/not_true', contract: contract, object: contact, field_name: 'valid'
```
and then run the tests to see that, for example, they fail.

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid (hunk 30 above)
Comment (fklosowski) [reply app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: Well, I expected it won't be clear enough. But it is quite simple: Only the last test's debug messages will be visible in the final output. Previous tests output is swallowed by the last test's output.

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid (hunk 16 above)
Comment (fklosowski) [reply app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: Applied this text change.

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid (hunk 2 above)
Comment (fklosowski) [reply app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: Linked :)

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid (hunk 4 above)
Comment (fklosowski) [reply app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: Done:)

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid (hunk 5 above)
Comment (fklosowski) [reply app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: Ok, applied this suggestion, thank you  :)

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid (hunk 30 above)
Comment (simplykatsa) [reply app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: Oh, I see, and this was my guess as well.

How about phrashing it something like this? But please modify if you feel that I altered the meaning :)

_You’ll only see the debug output in the browser because this test is the last one in the file. Any debug output from earlier tests gets overwritten and won’t appear in the final rendered output._

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid (hunk 20 above)
Comment (simplykatsa) [reply app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: Now it is clear. Can you add this above to the documentation? Or just a hint, like modify input data in the assign data lines or switch between 'function contract = 'modules/tests/assertions/not_true' and 'modules/tests/assertions/true'

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid (hunk 3 above)
Comment (fklosowski) [reply app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: I changed this :)

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid (hunk 30 above)
Comment (fklosowski) [reply app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: Ok, I checked it one more time and I think we should remove this info. Looks like liquid is able to render consecutive outputs. Previous tests output is NOT swallowed by the next test's output. My mistake :/

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid (hunk 20 above)
Comment (fklosowski) [reply app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: Ok, I extende this part:)

File: app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid (hunk 3 above)
Comment (fklosowski) [reply app/views/pages/get-started/contact-us-tutorial/testing-contact-create-command.liquid]: Fixed:)

**Review by simplykatsa — APPROVED** [review]
